src/hotspot/cpu/x86/vm_version_x86.hpp

```diff
@@ -524,36 +524,49 @@ class VM_Version : public Abstract_VM_Version {
     int          zmm_save[16*4]; // Save zmm0, zmm7, zmm8, zmm31
   };
 
+  class CpuidInfo : public _CpuidInfo {
```

Comment:
Why not just declare the original `CpuidInfo` as a class instead of extending the struct ???

Reply:
This way the members are always zero-initialized. Which simplifies existing code
```
-VM_Version::CpuidInfo VM_Version::_cpuid_info = { 0, };
+VM_Version::CpuidInfo VM_Version::_cpuid_info;
```
And makes it more foolproof - therefore also fixing an existing bug (I haven't been aware of so far) in my code in the [CRaC](https://openjdk.org/projects/crac/) branch.

Although the ` = { 0, }` initialization above was not needed as it is a static member anyway.

That means I should change the name of this patch as it is no longer just a "refactor" ization.

Or should I stop the zero-initialization?

Reply:
Sorry I'm not familiar enough with C++ initialization rules for structs versus classes here. IIUC:
- the original  struct instance was explicitly zero-initialized. 
- If we just made it a class there would not be any zero initialization
- If the class extends the struct then we (somehow) regain zero initialization
??

Maybe methods on a struct weren't so bad after all. I'd like to hear other views on this code change. (Just to be clear 2 reviews are needed for hotspot code changes anyway.)

Thanks

Reply:
>     * the original  struct instance was explicitly zero-initialized.

It had to be zero-initialized by user of the struct. Which is error prone if you forget the zero-initialization during each use of the struct type.

>     * If we just made it a class there would not be any zero initialization

struct vs. class does not matter.

>     * If the class extends the struct then we (somehow) regain zero initialization
>       ??

zero-initialized:
```
echo 'struct A { long a; }; void f() { A a=A(); }'|g++ -c -o 1.o -Wall -x c++ -;objdump -dC 1.o
```
NOT zero-initialized:
```
echo 'struct A { long a; }; void f() { A a    ; }'|g++ -c -o 1.o -Wall -x c++ -;objdump -dC 1.o
```
It is important if one calls the default constructor or not. With one inheritance one can enforce the zero-initialization by explicitly calling the inner `_CpuidInfo` default-constructor from the outer class `CpuidInfo`.

> Maybe methods on a struct weren't so bad after all. I'd like to hear other views on this code change. (Just to be clear 2 reviews are needed for hotspot code changes anyway.)

struct vs. class does not matter here. I am just mixing two refactorizations together. I can remove the zero-initialization (I can submit it as a separate merge request but then I do not need that part myself).